feat: add support for run task results callback

karvounis-form3 opened the pull request:
## Description

This PR aims to address the missing support for Run Tasks Integration API. I have already raised an issue https://github.com/hashicorp/go-tfe/issues/862 about this.
It introduces the following:
- RunTaskRequest struct which can be used to objectify the request you get from a TFC/E run task
- RunTasksCallback interface which can be used to update the run task with status and messages
  - TaskResultsCallbackOptions defines the attributes required in order to perform the update

## Testing plan

Unfortunately, it is not easy to test the Update function because you first need a running server which listens for incoming TFC/E requests. Then, you need a workspace run which is configured to use that run task endpoint.

However, I added the following unit tests and their output when run locally:

- TestTaskResultsCallbackRequestOptions_Validate

![Screenshot 2024-06-19 at 8 27 28 AM](https://github.com/hashicorp/go-tfe/assets/91056119/eb2c32af-9afd-4c45-9ed3-b637101de590)

- TestTaskResultsCallbackRequestOptions_Marshal

![Screenshot 2024-06-19 at 8 28 01 AM](https://github.com/hashicorp/go-tfe/assets/91056119/b10474ba-382e-48c9-a973-2c984827ccbd)

- TestTaskResultsCallbackUpdate_Validate

![Screenshot 2024-06-19 at 8 28 25 AM](https://github.com/hashicorp/go-tfe/assets/91056119/a5fcf567-c60c-4328-b69f-b07817681c1d)

- TestTaskResultsCallbackUpdate

![Screenshot 2024-06-19 at 8 28 52 AM](https://github.com/hashicorp/go-tfe/assets/91056119/b1248c1e-6a04-48ec-b044-4c6d85972b90)

## External links

https://github.com/hashicorp/terraform-run-task-scaffolding-go this HashiCorp repo contains part of the code that can be found here.

sebasslash commented on run_task_request.go:
```diff
@@ -0,0 +1,53 @@
+	"time"
+)
+
+const (
+	// RunTaskRegistrationRunId is the string that TFC/E sends when it just wants to register the Run Task.
+	RunTaskRegistrationRunId = "run-xxxxxxxxxxxxxxxx"
+)
```
Would like to understand this more, along with the `IsEndpointValidation` check. Is the intent to ensure `r.RunID` is a valid run external ID (in the format of `RunTaskRegistrationRunID`)? As I read it, `IsEndpointValidation` will only return true if `r.RunID == "run-xxxxxxxxxxxx"` which should never be the case. 

If we do want to ensure that it's in proper external ID format, use the helper [validStringID()](https://github.com/hashicorp/go-tfe/blob/193d7a6b5066d8f4528ae0a38da39e7c82360761/validations.go#L29)

sebasslash reviewed (changes requested):
Awesome, thanks for submitting this PR & sorry for the delay in reviewing. Some things to chew on in the meantime ⬇️

karvounis-form3 commented on run_task_results_callback.go:
```diff
@@ -0,0 +1,114 @@
+
+// TaskResultsCallbackOptions represents the options for a TFE Task result callback request
+// https://developer.hashicorp.com/terraform/enterprise/api-docs/run-tasks/run-tasks-integration#request-body-1
+type TaskResultsCallbackOptions struct {
+	Data *TaskResultsCallbackData `json:"data"`
+}
+
+type TaskResultsCallbackData struct {
+	// Required: Must be set to `task-results`
+	Type *string `json:"type"`
+	// Required: Attributes of the Task Results Callback Response
+	Attributes    *TaskResultsCallbackDataAttributes `json:"attributes"`
+	Relationships *TaskResultsCallbackRelationships  `json:"relationships,omitempty"`
+}
+
+type TaskResultsCallbackDataAttributes struct {
+	// Status Must be one of TaskFailed, TaskPassed or TaskRunning
+	Status TaskResultStatus `json:"status"`
+	// Message A short message describing the status of the task.
+	Message string `json:"message,omitempty"`
+	// URL that the user can use to get more information from the external service
+	URL string `json:"url,omitempty"`
+}
+
+type TaskResultsCallbackRelationships struct {
+	// Outcomes A run task result may optionally contain one or more detailed outcomes, which improves result visibility and content in the Terraform Cloud user interface.
+	// https://developer.hashicorp.com/terraform/enterprise/api-docs/run-tasks/run-tasks-integration#outcomes-payload-body
+	Outcomes *TaskResultsCallbackRelationshipsOutcomes `json:"outcomes"`
```
Hey @sebasslash thanks for the comment! I will work on that asap!

karvounis-form3 replied:
I made the changes and I have confirmed that they work! Thanks for the amazing suggestion. I learned something new!

sebasslash commented on run_task_results_callback.go:
```diff
@@ -0,0 +1,79 @@
+	return req.Do(ctx, nil)
+}
+
+// TaskResultCallbackRequestOptions represents the TFC/E Task result callback request
+// https://developer.hashicorp.com/terraform/enterprise/api-docs/run-tasks/run-tasks-integration#request-body-1
+type TaskResultCallbackRequestOptions struct {
+	Type     string               `jsonapi:"primary,task-results"`
+	Status   TaskResultStatus     `jsonapi:"attr,status"`
+	Message  string               `jsonapi:"attr,message,omitempty"`
+	URL      string               `jsonapi:"attr,url,omitempty"`
+	Outcomes []*TaskResultOutcome `jsonapi:"relation,outcomes,omitempty"`
+}
+
+// TaskResultOutcome represents a detailed TFC/E run task outcome, which improves result visibility and content in the TFC/E UI.
+// https://developer.hashicorp.com/terraform/enterprise/api-docs/run-tasks/run-tasks-integration#outcomes-payload-body
+type TaskResultOutcome struct {
+	Type        string                      `jsonapi:"primary,task-result-outcomes"`
+	OutcomeID   string                      `jsonapi:"attr,outcome-id,omitempty"`
+	Description string                      `jsonapi:"attr,description,omitempty"`
+	Body        string                      `jsonapi:"attr,body,omitempty"`
+	URL         string                      `jsonapi:"attr,url,omitempty"`
+	Tags        map[string][]*TaskResultTag `jsonapi:"attr,tags,omitempty"`
+}
```
Move these above the `Update()` call for visibility. The style ordering for a file is:
1. Interface definition
2. Struct definitions
3. Receiver method implementations.

karvounis-form3 replied:
I moved them up just before the `Update` function

sebasslash reviewed (changes requested):
So the implementation look great ✅ However I do think we ought to add some unit tests around this behavior, particularly ensuring that things are being serialized correctly (`TaskResultTag` gives me some pause given mix-&-matching of `json/jsonapi` struct tags).

Writing a true integration test would not be plausible here, as it would require the test function to expose a publicly available server. 

What sort of unit tests might be possible here?
- Test for validation errors, scenarios that would result in `ErrInvalidTaskResultsCallbackStatus`, `ErrInvalidCallbackURL`, `ErrInvalidAccessToken` and `ErrInvalidTaskResultsCallbackType`. 
- Ensure the `Authorization` header is set to the correct token provided in the request sent from TFC. In the test function, you can spin up an [httptest.Server](https://pkg.go.dev/net/http/httptest#example-Server), define an endpoint that validates the header is the correct token, and set that endpoint URL to `TaskResultCallbackURL` in a mocked `RunTaskRequest`. 
- Ensure `TaskResultTag` is serialized correctly.

Last thing, make sure you define the server in `helper_test.go`. An example signature can be: 
```
func runTaskCallbackMockServer(t *testing.T) *httptest.Server {}
```

Once you've added tests, make sure to run them locally and add the test output to the PR description. CI won't fire given this is a fork's branch, so once everything is a-ok I will rebase this PR onto a local branch to trigger CI.

karvounis-form3 commented:
@sebasslash thank you for your suggestions to improve the code. I added the unit tests and added screenshots from running the tests locally in the PR body as requested. Can you please review the PR once more?

karvounis-form3 commented:
Fixed code based on `golangci-lint` [suggestions](https://github.com/hashicorp/go-tfe/actions/runs/9576670518/job/26657617891?pr=867)

karvounis-form3 commented:
@Maed223 maybe you can rerun the CI?

karvounis-form3 commented:
@Maed223 https://github.com/hashicorp/go-tfe/actions/runs/9664123031/job/26672604714#step:2:1 I do not understand what actions to take to fix that issue

karvounis-form3 commented:
@sebasslash @Maed223 any thoughts on this?

sebasslash commented:
Since this PR comes from a fork, I will rebase your changes onto a local branch so that your changes execute against CI.

karvounis-form3 commented:
@sebasslash can we please prioritise merging that PR? We need that code

sebasslash commented:
Closing in favor of #929
